packages/web/app/src/pages/target-laboratory.tsx

```diff
@@ -655,3 +692,108 @@ function useApiTabValueState(graphqlEndpointUrl: string | null) {
+    const consoleEl = consoleRef.current;
+    consoleEl?.scroll({ top: consoleEl.scrollHeight, behavior: 'smooth' });
+  }, [props.logs, isOpen]);
+
+  const logColor = {
+    error: 'text-red-400',
+    info: 'text-emerald-400',
+    warn: 'text-yellow-400',
+    log: '', // default
+  };
+
+  return (
+    <Collapsible
+      open={isOpen}
+      onOpenChange={setIsOpen}
+      className={cn('flex max-h-[200px] w-full flex-col overflow-hidden bg-[#030711]')}
+      id="preflight-script-logs"
+    >
+      <div
+        className={cn(
+          'flex shrink-0 items-center justify-between px-4 py-3',
+          isOpen ? 'border-b' : 'border-b-0',
+        )}
+      >
+        <CollapsibleTrigger asChild>
+          <Button
+            variant="ghost"
+            size="sm"
+            className="flex h-auto items-center gap-2 p-0 hover:bg-transparent"
+            data-cy="trigger"
+          >
+            <ChevronDownIcon
+              className={`size-4 text-gray-500 transition-transform ${
+                isOpen ? 'rotate-0' : '-rotate-90'
+              }`}
+            />
+            <h2 className="text-[15px] font-normal">Preflight Script Logs</h2>
+          </Button>
+        </CollapsibleTrigger>
+        <div className="flex items-center gap-2">
+          <Button
+            variant="ghost"
+            size="icon"
+            data-cy="erase-logs"
+            className={cn(
+              'size-8 text-gray-500 hover:text-white',
+              isOpen ? 'visible' : 'invisible',
+            )}
+            onClick={props.onClear}
+          >
+            <EraserIcon className="size-4" />
+            <span className="sr-only">Clear logs</span>
+          </Button>
+        </div>
+      </div>
+      <CollapsibleContent
+        className="grow overflow-auto p-4 font-mono text-xs/[18px]"
+        ref={consoleRef}
+        data-cy="logs"
+      >
+        {props.logs.length === 0 ? (
+          <div
+            data-cy="empty-state"
+            className="flex flex-col items-center justify-center text-gray-400"
+          >
+            <p>No logs available</p>
+            <p>Execute a query to see logs</p>
+          </div>
+        ) : (
+          <>
+            {props.logs.map((log, index) => {
+              if (typeof log !== 'string' && 'type' in log && log.type === 'separator') {
+                return <hr key={index} className="my-2 border-dashed border-current" />;
+              }
+
+              let logType: 'error' | 'warn' | 'info' | 'log' = 'log';
+              let logMessage = '';
+
+              if (log instanceof Error) {
+                logType = 'error';
+                logMessage = `${log.name}: ${log.message}`;
+              } else if (typeof log === 'string') {
+                logType = log.split(':')[0].toLowerCase() as 'error' | 'warn' | 'info' | 'log';
+                logMessage = log.substring(log.indexOf(':') + 1).trim();
+              } else {
+                return null;
```

Comment:
maybe json stringify and spit out a warning in this case? And/or send to sentry since I dont think this should ever happen.

Reply:
True, I wanted to turn those string-based lines into structured format in the following PR, but we can log a warning here now